### ISSUE TITLE
Schema: support for TextDocumentContentProvider URIs

### ISSUE THREAD

**Issue by rewrking** [redhat-developer/vscode-yaml#986]
### Is your enhancement related to a problem? Please describe.
VS Code has the ability for extension developers to generate a JSON schema dynamically using a custom URI, such as `my-extension:///my-json-schema`, although the API is just to provide dynamic text content. See here:
  https://code.visualstudio.com/api/extension-guides/virtual-documents#textdocumentcontentprovider

A real world example of this would be if the text content for the json schema was provided by an executable that might change versions frequently. The TextDocumentContentProvider always provides the extension with the schema used by the current executable found in Path.

I tested this out with two scenarios:

```json
"yaml.schemas": {
  "my-extension:///my-json-schema": [
    "myfile.yaml"
  ]
}
```

and in my extension:

```json
"yamlValidation": [
  {
    "fileMatch": [ "myfile.yaml" ],
    "url": "my-extension:///my-json-schema"
  }
]
```

Both had the same result, where an error in the first character of the yaml file displays with:
`Unable to load schema from 'my-extension:///my-json-schema': No content. | YAML(768)`

### Describe the solution you would like
This has been in the VS Code extension API for at least a few years, so I expected this to just work. I'm not sure what needs to be updated in this extension for it to resolve those URIs correctly.

### Describe alternatives you have considered
This is a definitely a nice to have. The alternative for me is to use what's available now - provide a static schema from a URL or file path, but it could easily go out of sync with the executable I'm designing the schema for.

**Comment by gorkem** [comment]
This sounds like a good idea. I am not sure when we can get to it though.

**Comment by vaclavHala** [comment]
Stumbled into this issue when I was looking for similar thing. I believe your requirement can be satisfied using the current vscode-yaml API already.

In package.json define your custom URI as you showed
```json
"yamlValidation": [
  {
    "fileMatch": [ "myfile.yaml" ],
    "url": "my-extension:///my-json-schema"
  }
]
``` 

then in code of your extension activation
```ts
// https://github.com/redhat-developer/vscode-yaml/blob/main/src/schema-extension-api.ts
const rhYamlApi: ExtensionApi = vscode.extensions...
rhYamlApi.registerContributor(
    'my-extension',
    () => undefined,
    async (rawUri: string) => {
        // you get "my-extension:///my-json-schema" as arg here so you can use rest of the URI 
        // to pass additional info to this loader, no need to use different scheme for each "type" of schema you work with
        const uri = vscode.Uri.parse(rawUri)
        const doc = await vscode.workspace.openTextDocument(uri)
        return doc.getText()
    }
  ): boolean;
```
The trick is using scheme of your custom uri (`my-extension` here) as name of your custom contribution so this code triggers https://github.com/redhat-developer/vscode-yaml/blob/main/src/schema-extension-api.ts#L149 . I did not find this documented anywhere, I discovered it just by reading the source so I hope this is the way it is intended to work and not something that will change in the future (@gorkem  ? )

On one hand this is not exactly what you wanted - out of the box support for `TextDocumentContentProvider`, on the other if what you really want is to have custom logic for "getting schema content" where you can call arbitrary code, e.g. spawn external program, this should do quite fine

**Comment by rewrking** [comment]
That worked perfectly! Thanks for the detailed response. The only thing I had to add was:

```ts
if (!rhYaml.isActive) await rhYaml.activate();
```

... before calling `registerContributor`. I also had to check if rhYaml is defined of course (in case someone doesn't have this extension installed).  Really appreciate the response! Assuming `registerContributor` itself is stable, this issue is good on my end.
